server/src/main/java/org/elasticsearch/common/bytes/PagedBytesReference.java

```diff
@@ -48,6 +49,29 @@ public BytesReference slice(int from, int length) {
         return new PagedBytesReference(byteArray, offset + from, length);
     }
 
+    @Override
+    public BytesReference copy(int from, int length) {
+        Objects.checkFromIndexSize(from, length, this.length);
+        final int offsetFirstPage = (Math.addExact(offset, from)) % PAGE_SIZE;
+        // adjust offset and length for doing aligned page copies
+        final int adjustedFrom = from - offsetFirstPage;
+        final int adjustedLength = Math.addExact(length, offsetFirstPage);
+        final ByteArray byteArray = BigArrays.NON_RECYCLING_INSTANCE.newByteArray(adjustedLength);
+        final long offset = this.offset + adjustedFrom;
+        final BytesRef slice = new BytesRef();
+        int nextFragmentSize = Math.min(adjustedLength, PAGE_SIZE);
+        int position = 0;
+        while (nextFragmentSize != 0) {
+            final boolean materialized = this.byteArray.get(offset + position, nextFragmentSize, slice);
```

Comment:
Can you use the `iterator()` method which already checks for the materialization?

Reply:
The iterator does not let you start from a specific page so you would need to iterate until you find the right page to start from?

Reply:
You should be able to do `.slice(from, length).iterator()` and use it to massively simplify this function. You'd simply read page slices, and to byteArray.set() to copy each page. Right?

Reply:
Nice, I implemented that way, looks much better.